src/iguana/algorithms/clas12/FiducialFilter/Algorithm.cc

```diff
@@ -0,0 +1,294 @@
+    else if(pid==211 || pid==-211 || pid==2212){
+        if(torus<0) return DC_fiducial_cut_theta_phi_pass1(traj_row, torus, pid);
+        else if(torus>0) return DC_fiducial_cut_XY_pass1(traj_row, torus, pid);
+        else return true;
+    }
+    return true;
+  }
+
+    
+
+  bool FiducialFilter::DC_fiducial_cut_XY_pass1(FiducialFilter::traj_row_data const traj_row, float const torus, int const pid) const
+  {
+    
+      const auto minparams = ((torus<0) ? minparams_in_XY_pass1 : minparams_out_XY_pass1);
+      const auto maxparams = ((torus<0) ? maxparams_in_XY_pass1 : maxparams_out_XY_pass1);
+      double X=0;
+      double Y=0;
+      for(int r = 0 ; r < 3; r++){
+        X=0;
+        Y=0;
+        switch(r){
+        case 0:
+          X = traj_row.x1;
+          Y = traj_row.y1;
+          break;
+        case 1:
+          X = traj_row.x2;
+          Y = traj_row.y2;
+          break;
+        case 2:
+          X = traj_row.x3;
+          Y = traj_row.y3;
+          break;
+        }
+
+        int sector = traj_row.sector;
+
+        if(sector == 2)
+          {
+        const double X_new = X * std::cos(-60 * PI / 180) - Y * std::sin(-60 * PI / 180);
+        Y = X * std::sin(-60 * PI / 180) + Y * std::cos(-60 * PI / 180);
+        X = X_new;
+          }
+
+        if(sector == 3)
+          {
+        const double X_new = X * std::cos(-120 * PI / 180) - Y * std::sin(-120 * PI / 180);
+        Y = X * std::sin(-120 * PI / 180) + Y * std::cos(-120 * PI / 180);
+        X = X_new;
+          }
+
+        if(sector == 4)
+          {
+        const double X_new = X * std::cos(-180 * PI / 180) - Y * std::sin(-180 * PI / 180);
+        Y = X * std::sin(-180 * PI / 180) + Y * std::cos(-180 * PI / 180);
+        X = X_new;
+          }
+
+        if(sector == 5)
+          {
+        const double X_new = X * std::cos(120 * PI / 180) - Y * std::sin(120 * PI / 180);
+        Y = X * std::sin(120 * PI / 180) + Y * std::cos(120 * PI / 180);
+        X = X_new;
+          }
+
+        if(sector == 6)
+          {
+        const double X_new = X * std::cos(60 * PI / 180) - Y * std::sin(60 * PI / 180);
+        Y = X * std::sin(60 * PI / 180) + Y * std::cos(60 * PI / 180);
+        X = X_new;
+          }
+
+
+        int this_pid = 0;
+
+        switch (pid)
+          {
+          case 11: 
+        this_pid = 0; 
+        break;
+          case 2212: 
+        this_pid = 1; 
+        break;
+          case 211: 
+        this_pid = 2; 
+        break;
+          case -211: 
+        this_pid = 3; 
+        break;
+          case 321: 
+        this_pid = 4; 
+        break;
+          case -321: 
+        this_pid = 5; 
+        break;
+          default: 
+        return false; 
+        break;
+          }
+        double calc_min = minparams[this_pid][sector - 1][r][0] + minparams[this_pid][sector - 1][r][1] * X;
+        double calc_max = maxparams[this_pid][sector - 1][r][0] + maxparams[this_pid][sector - 1][r][1] * X;
```

Comment:
If `sector==0`, this will be undefined behavior; I believe this is the issue that `ubsan` is complaining about. I'm not sure how to best handle this.

Reply:
Addressed in bab5dc8

Reply:
`ubsan` still complains

Reply:
you need to make sure there is _no possible way_ for `sector<0` by the time it's used as an array index.

Reply:
I see the problem was in the definition of the default struct itself. Seems to be fixed now in 1c380c954109efe16c1c2cdd57359d3fe5dd5745 however I am unsure how to comment the struct in `Algorithm.h` so that it passes the docs check.